### PR TITLE
Revert shared container rootfs

### LINKED ISSUE #13775: [1.7.0-rc*] Data written to ZFS mounts isn't visible in host mounts
This is a weird one. I'll keep digging into this and flesh out the issue as I discover more, but:

I have this directory. It's a ZFS dataset.

```
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# stat -f . | awk '/Type/ { print $NF }'
zfs
```

```
# zfs list | grep a929 | grep week
locutus/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01                                             47K   144G    28K  /u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01
locutus/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01/6ecddc3dd27d5645ec1a2153e7806f99fc1727e0    19K   144G    19K  /u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01/6ecddc3dd27d5645ec1a2153e7806f99fc1727e0
```

When I mount it into a container, the behaviour is quite different on 1.6.0 and 1.7.0, probably due to the Volumes refactor:

##### 1.6.0:

```
root@0beae52c57ab:/app# grep zfs /proc/mounts
locutus/shopify-f744a0801011b8a849807d3b53df91e719e1bd5ad8523015d984be663482480f/weekof-2015-06-01 /test zfs rw,relatime,xattr,noacl 0 0
locutus/shopify-f744a0801011b8a849807d3b53df91e719e1bd5ad8523015d984be663482480f/weekof-2015-06-01/03f195f2b7f552b7fa8093364fcb4fc0727b574a /test/03f195f2b7f552b7fa8093364fcb4fc0727b574a zfs rw,relatime,xattr,noacl 0 0
root@0beae52c57ab:/app#
```

Note that here, we see nested volumes.

##### 1.7.0:

For some volumes, depending on circumstances I haven't yet been able to figure out, we get this:

```
# grep zfs /proc/mounts
locutus/spy-499d8dac8bb0e5d88a981afb20e1fd26fac699799b72cef71f365376e79e89c4 /test zfs rw,relatime,xattr,noacl 0 0
locutus/spy-499d8dac8bb0e5d88a981afb20e1fd26fac699799b72cef71f365376e79e89c4/weekof-2015-06-01 /test/weekof-2015-06-01 zfs rw,relatime,xattr,noacl 0 0
```

i.e.: the same, and there is no issue.

In other cases, and again, no idea what causes this:

```
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# docker run -it --entrypoint=/bin/bash -v $(pwd):/test:rw ba5e -l
root@1469d291f2af:/app# grep zfs /proc/mounts
locutus /test zfs rw,relatime,xattr,noacl 0 0
root@1469d291f2af:/app#
```

So note, here, that this directory is mounted on the host as a ZFS dataset, with the same behaviour on my 1.6.0 nodes and my 1.7.0 nodes. However, the 1.7.0 nodes sometimes bind it inside the containers as `locutus` (my zpool name) rather than the full dataset name.

There is some very interesting behaviour as a result of this:

```
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# ls
6ecddc3dd27d5645ec1a2153e7806f99fc1727e0
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# docker run -it --entrypoint=/bin/bash -v $(pwd):/test:rw ba5e -l
root@bde4abe26612:/app# touch /test/what
root@bde4abe26612:/app# logout
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# ls
6ecddc3dd27d5645ec1a2153e7806f99fc1727e0
root@borg2:/u/apps/locutus/shared/zfs/shopify-a929cf7d522b5f094635fadbec2fbf96fc973c96764e35303e4bc80dc6974dea/weekof-2015-06-01# docker run -it --entrypoint=/bin/bash -v $(pwd):/test:rw ba5e -l
root@2ee6bbc4d492:/app# ls /test
6ecddc3dd27d5645ec1a2153e7806f99fc1727e0  what  wip-6ecddc3dd27d5645ec1a2153e7806f99fc1727e0.10704
root@2ee6bbc4d492:/app#
```

To summarize what's happened above:

1. List files in dataset on host. Get `ds/{a}`.
2. Mount dataset in new container. See `ds/{a}`. Create `ds/{b}` Now container sees `ds/{a,b}`.
3. Exit container. List files on host. See `ds/{a}`.
4. Start another container with the same mount. See `ds/{a,b}`.

What I would expect to see is that the writes from (2) are visible on the host in step 3. I think this is somehow very related to the incorrect-looking mount name in `/proc/mounts`. I would normally blame ZFS for this, but we have strong empirical evidence that this happens only on docker 1.7.0.

/cc @calavera

### PR DESCRIPTION
**Pull request by crosbymichael**
This is breaking various setups where the host's rootfs is mount shared
correctly and breaks live migration with bind mounts.

Fixes #13775

Signed-off-by: Michael Crosby crosbymichael@gmail.com

### CODE REVIEW

**Comment by icecrime** [comment]
@rhatdan @vbatts We're gonna take this in for 1.7.0 instead of reverting docker/libcontainer#520, but you're probably going to have to take this into account.

**Comment by icecrime** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by vbatts** [comment]
hold on. he has not isolated it to this

**Comment by vbatts** [comment]
hold on. he has not isolated it to this

**Comment by rhatdan** [comment]
We disagree with this.  This is a needed feature, and you are reverting it for an unknown, potentially kernel issue.

**Comment by calavera** [comment]
@vbatts I think @crosbymichael found other broken cases due to making the root slave. We should keep the old behavior until we can ensure that slave works 100% of the times.

**Comment by rhatdan** [comment]
Can't we make this a daemon flag, and then set allow admins to set it.

**Comment by icecrime** [comment]
@rhatdan We're not contesting the feature in itself, we're just saying that we can't ship 1.7.0 with the new default, and right now we're blocked by this.

**Comment by calavera** [comment]
@rhatdan we can do that, or even try to detect it, like #13849 did, but we need to keep the option that's stable for now.

**Comment by jfrazelle** [comment]
in 1.8 we can have a daemon flag possibly (mo flags mo problems) but for 1.7 we are going back to original behavior we are not going to break 90% of kernels for some new thing

**Comment by crosbymichael** [comment]
With this we get errors when we don't have this set causing mount sharing issues.

`Error (mount.c:681): Mount 197 (master_id: 1 shared_id: 0) has unreachable sharing`

**Comment by vbatts** [comment]
... i'm not seeing where 90% of kernels are broken...

**Comment by rhatdan** [comment]
We have seen no breakage in Fedora Rawhide docker-1.7 release, and most of our teams are using docker-1.7.

**Comment by jfrazelle** [comment]
"the norm" is what I was referring to

**Comment by jfrazelle** [comment]
as per @rhatdan comment here
https://github.com/docker/docker/issues/13775#issuecomment-110452144

> "I believe the opposite is true for regular kernels."

**Comment by rhatdan** [comment]
The default for the kernel is file systems with private.  On systemd based systemd the default is changed to rshared.   

rprivate surpises Admins.  In that if you setup a file system mount namespace, most users would expect mounts on the host to show up in the container, but they don't unless / is rshared with its children.

We have been running rhel/Fedora kernels like this since before RHEL6 arrived.   Probably back to Fedora 4 or 5.

**Comment by icecrime** [comment]
The current situation is that we introduced a regression that was reported in #13775, and noticed by @crosbymichael in his CRIU experiments. It's blocking the 1.7.0 release process, and we need to move forward, which is why I'm merging this.

That being said, I understand how mounts not propagating can be surprising, and we probably shouldn't stop there. I'm gonna be honest, I don't see many use cases for it, but if you want to submit a PR adding a daemon flag to offer that choice, I think we'd take it.

So to sum up, this PR is not us disagreeing with you, it's just us being in a code freeze and having a regression that needs fixing :-/ Sorry.

**Comment by jfrazelle** [comment]
cherry-picked

**Comment by rhatdan** [comment]
The biggest use case this breaks is using autofs on the host to mount directories into containers.

**Comment by rhatdan** [comment]
But other use cases like using SPC container on the host where you have / mounted into the container, if mounts on the host don't show up in this environment, then users see breakage.

**Comment by rhatdan** [comment]
CRIU issue would seem to be breakage in the kernel,  I don't know what CRIU would care about the way the shareing is setup for mount points.

**Comment by SamYaple** [comment]
This revert broke sharing new namespaces from host to container. We use (used?) that in Kolla to split all of the Openstack Neutron components into separate containers.

We would love to see a way to turn off private namespaces, even if the default remains on. A daemon argument would be fine with me.

**Comment by pmorie** [comment]
This also breaks running the kubernetes kubelet and the OpenShift node controller (or all-in-one) binary in a container.

It would be great if we could control the propagation mode of the mount with a flag on the bindmount spec a la `:Z` .
